Allow use of alternative database interfaces

tsvallender opened the pull request:
### Motivation / Background

Currently running `rails dbconsole` is hardcoded to use the default interface for the given database (e.g. `psql` for Postgres). Other interfaces exist however, such as `pgcli` which I’m a fan of.

### Detail

Allows the use of alternative interfaces by specifying the `cmd` key in the database configuration.

### Additional information

Considered the use of an environment variable for this, but this approach seemed more flexible; and environment variable can still be used via ERB.

### Checklist

Before submitting the PR make sure the following are checked:

* [x] This Pull Request is related to one change. Unrelated changes should be opened in separate PRs.
* [x] Commit message has a detailed description of what changed and why. If this PR fixes a related issue include it in the commit message. Ex: `[Fix #issue-number]`
* [x] Tests are added or updated if you fix a bug or add a feature.
* [x] CHANGELOG files are updated for the changed libraries if there is a behavior change or additional feature. Minor bug fixes and documentation changes should not be included.

eileencodes commented on guides/source/configuring.md:
````diff
@@ -3231,6 +3231,13 @@ configured without modifying the URL set in the deployment environment. See:
+
+```yaml
+development:
+  cmd: pgcli
````
I don't like `cmd`. It's not clear enough. How about `database_cli`? I'm also not sure this should be per-database. Is there any use case where one db would want to use `pgsql` and another `pgcli`?

tsvallender replied:
Agreed on the first point. I thought it might be common to want to use something like `pgcli` locally but not want to install that on your production servers. But of course it’s easy enough to just define it in `default:` and optionally override, unless I’m misunderstanding and you mean it should be defined somewhere else entirely?

eileencodes replied:
I don't mean to define it in `default`, I mean maybe it should be an attr_accessor on ActiveRecord like `schema_cache_ignored_tables`. IE `ActiveRecord.database_cli = :pgcli` or something like that. It could even support a hash of values if we considered you might be using mysql and postgres in a single app (`ActiveRecord.database_cli = { postgresql: :pgcli, mysql: :mysql_cli }`

I think this example would be pretty super rare and error prone, I'm not sure it should be supported via the database config:

```yaml
development:
  primary:
    database: "my_db"
    adapter: "postgres"
    database_cli: "pgsql"
  secondary:
    database: "my_other_db"
    adapter: "postgres"
    database_cli: "pgcli"
```

tsvallender replied:
Okay, I think that makes sense. Just to be sure I understand before I take a shot at this tomorrow, you’re suggesting:
- Add a `database_cli` attr_accessor to ActiveRecord
- This would define the default CLI for each database adapter
- Users would override this with `config.active_record.database_cli = { postgres: "pgcli" }`

tsvallender replied:
I’ve changed the PR to work in that way, and I think the result is much nicer thanks. I’ve also squashed the commits ready to merge if you’re happy with it at this point.

Is there anywhere else this should be documented to encourage other database adapter authors to use the convention?

There is a test failure currently but I think it’s something intermittent as I’m not getting it locally and it only occurred after squashing, not any actual code change.

eileencodes commented on activerecord/CHANGELOG.md:
```diff
@@ -1,3 +1,7 @@
+*   Support use of alternative database interfaces via the `cmd` key in the database configuration file.
```
This should include an example of usage.

eileencodes commented on activerecord/lib/active_record/connection_adapters/abstract_adapter.rb:
```diff
@@ -91,7 +91,7 @@ def self.build_read_query_regexp(*parts) # :nodoc:
 
       def self.find_cmd_and_exec(commands, *args) # :doc:
-        commands = Array(commands)
+        commands = Array(commands).compact
```
Rather than needing to the first value is `nil`, there should be a default for each db set. This shouldn't have to call compact.

eileencodes reviewed (changes requested):
Thanks for the PR. I think the idea makes sense but I left some comments about the implementation. I don't like `cmd` as a key, and there we can avoid calling `compact` or asserting `nil` in the tests, instead there should be a default set for each db adapter.

Also can you check out the failing tests. Thanks!

tsvallender commented:
> Thanks for the PR. I think the idea makes sense but I left some comments about the implementation. I don't like `cmd` as a key, and there we can avoid calling `compact` or asserting `nil` in the tests, instead there should be a default set for each db adapter.

Thanks for reviewing! I believe I’ve responded to everything you said (and the code is better for it).

tsvallender committed:
Allow use of alternative database interfaces

The CLI tool used as a database interface is now specified and
customisable via ActiveRecord.database_cli.

This specifies the current defaults but allows them to be overridden by
users. It continues to accept array values to allow fallback options.

rafaelfranca commented on guides/source/configuring.md:
```diff
@@ -3230,6 +3230,14 @@ configured without modifying the URL set in the deployment environment. See:
 
 TIP: You don't have to update the database configurations manually. If you look at the options of the application generator, you will see that one of the options is named `--database`. This option allows you to choose an adapter from a list of the most used relational databases. You can even run the generator repeatedly: `cd .. && rails new blog --database=mysql`. When you confirm the overwriting ...
 
+By default, Rails will use the standard CLI interface for your database when running `rails dbconsole` (e.g. `psql` for PostgreSQL and `mysql` for MySQL). If you wish to use an alternative application, you can specify this with the `database_cli` ActiveRecord configuration option. To specify a fallback, you can use an array:
```
Instead of adding this here can we create a new section for this new config now like all the others?

https://github.com/rails/rails/blob/73608ff4110710d1dba81da9cd318440d42a9d23/guides/source/configuring.md#configuring-active-record

tsvallender replied:
That’s done—sorry I missed the obvious place :facepalm:
